### PR TITLE
chore: get testcontainers working with custom mix task

### PR DESCRIPTION
**Pull request by jarlah**
Run tests with testcontainers for elixir. Requires a bit of hacking. since commanded "needs"/"uses" its own database.
Im the maintainer of said testcontainers library so maybe this will trickle upstream ...

### CODE REVIEW

File: config/runtime.exs [hunk 1]
```diff
@@ -36,6 +36,12 @@ if config_env() == :prod do
     pool_size: String.to_integer(System.get_env("POOL_SIZE") || "10"),
     socket_options: maybe_ipv6
 
+  config :journi_plan, JourniPlan.EventStore,
+    # ssl: true,
+    url: database_url,
```
Comment (jarlah) [inline config/runtime.exs:41]: should propably use different database url for commanded eventstore

**Comment by jarlah** [comment]
- [ ] might not even need to use loki .. can call System.cmd directly

**Comment by jarlah** [comment]
we can see it succeeds by this cryptic tuple returned by loki:


```
{"==> elixir_uuid\nCompiling 1 file (.ex)\nGenerated elixir_uuid app\n==> gen_stage\nCompiling 10 files (.ex)\nGenerated gen_stage app\n==> floki\nCompiling 2 files (.erl)\nCompiling 29 files (.ex)\nGenerated floki app\n==> loki\nCompiling 7 files (.ex)\nGenerated loki app\n==> decimal\nCompiling 4 files (.ex)\nGenerated decimal app\n==> mime\nCompiling 1 file (.ex)\nGenerated mime app\n==> nimble_options\nCompiling 3 files (.ex)\nGenerated nimble_options app\n===> Analyzing applications...\n===> Compiling unicode_util_compat\n===> Analyzing applications...\n===> Compiling idna\n===> Analyzing applications...\n===> Compiling telemetry\n==> telemetry_metrics\nCompiling 7 files (.ex)\nGenerated telemetry_metrics app\n===> Analyzing applications...\n===> Compiling telemetry_poller\n==> telemetry_registry\nCompiling 1 file (.erl)\nCompiling 1 file (.ex)\nGenerated telemetry_registry app\n==> thousand_island\nCompiling 16 files (.ex)\nGenerated thousand_island app\n==> jason\nCompiling 10 files (.ex)\nGenerated jason app\n==> comeonin\nCompiling 3 files (.ex)\nGenerated comeonin app\n==> phoenix_html\nCompiling 6 files (.ex)\nGenerated phoenix_html app\n==> phoenix_template\nCompiling 4 files (.ex)\nGenerated phoenix_template app\n==> db_connection\nCompiling 17 files (.ex)\nGenerated db_connection app\n==> expo\nCompiling 2 files (.erl)\nCompiling 22 files (.ex)\nGenerated expo app\n==> phoenix_pubsub\nCompiling 11 files (.ex)\nGenerated phoenix_pubsub app\n==> plug_crypto\nCompiling 5 files (.ex)\nGenerated plug_crypto app\n==> hpax\nCompiling 4 files (.ex)\nGenerated hpax app\n===> Analyzing applications...\n===> Compiling mimerl\n==> dns_cluster\nCompiling 1 file (.ex)\nGenerated dns_cluster app\n==> gettext\nCompiling 18 files (.ex)\nGenerated gettext app\n==> fsm\nCompiling 1 file (.ex)\nGenerated fsm app\n==> ssl_verify_fun\nCompiling 7 files (.erl)\nGenerated ssl_verify_fun app\n==> exconstructor\nCompiling 1 file (.ex)\nGenerated exconstructor app\n===> Analyzing applications...\n===> Compiling backoff\n===> Analyzing applications...\n===> Compiling certifi\n==> ecto\nCompiling 56 files (.ex)\nGenerated ecto app\n==> plug\nCompiling 1 file (.erl)\nCompiling 40 files (.ex)\nGenerated plug app\n==> postgrex\nCompiling 68 files (.ex)\nGenerated postgrex app\n==> eventstore\nCompiling 60 files (.ex)\nGenerated eventstore app\n===> Analyzing applications...\n===> Compiling parse_trans\n==> ecto_sql\nCompiling 25 files (.ex)\nGenerated ecto_sql app\n==> nimble_pool\nCompiling 2 files (.ex)\nGenerated nimble_pool app\n===> Analyzing applications...\n===> Compiling metrics\n===> Analyzing applications...\n===> Compiling hackney\n==> commanded\nCompiling 67 files (.ex)\nGenerated commanded app\n==> commanded_eventstore_adapter\nCompiling 2 files (.ex)\nGenerated commanded_eventstore_adapter app\n==> commanded_ecto_projections\nCompiling 1 file (.ex)\nGenerated commanded_ecto_projections app\n==> castore\nCompiling 1 file (.ex)\nGenerated castore app\n==> esbuild\nCompiling 4 files (.ex)\nGenerated esbuild app\n==> tailwind\nCompiling 3 files (.ex)\nGenerated tailwind app\n==> mint\nCompiling 1 file (.erl)\nCompiling 20 files (.ex)\nGenerated mint app\n==> finch\nCompiling 14 files (.ex)\nGenerated finch app\n==> tesla\nCompiling 38 files (.ex)\nGenerated tesla app\n==> ex_docker_engine_api\nCompiling 221 files (.ex)\nGenerated ex_docker_engine_api app\n==> testcontainers\nCompiling 30 files (.ex)\nGenerated testcontainers app\n==> elixir_make\nCompiling 8 files (.ex)\nGenerated elixir_make app\n==> bcrypt_elixir\nmkdir -p \"/home/runner/work/JourniPlan/JourniPlan/_build/test/lib/bcrypt_elixir/priv\"\ncc -g -O3 -Wall -Wno-format-truncation -I\"/opt/hostedtoolcache/otp/ubuntu-22.04/OTP-25.3.2.14/x64/erts-13.2.2.10/include\" -Ic_src -fPIC -shared  c_src/bcrypt_nif.c c_src/blowfish.c -o \"/home/runner/work/JourniPlan/JourniPlan/_build/test/lib/bcrypt_elixir/priv/bcrypt_nif.so\"\nCompiling 3 files (.ex)\nGenerated bcrypt_elixir app\n==> websock\nCompiling 1 file (.ex)\nGenerated websock app\n==> bandit\nCompiling 50 files (.ex)\nGenerated bandit app\n==> swoosh\nCompiling 51 files (.ex)\nGenerated swoosh app\n==> websock_adapter\nCompiling 4 files (.ex)\nGenerated websock_adapter app\n==> phoenix\nCompil" <> ...,
 0}
 ```
 
 the zero atthe end ...

File: lib/mix/tasks/testcontainers_test.ex [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    # Start Testcontainers
+    {:ok, _} = Testcontainers.start_link()
+
+    # Configure and start a Postgres container
+    config = PostgresContainer.new()
+    {:ok, container} = Testcontainers.start_container(config)
+
+    # Get the connection URL
+    app_database_url = "postgres://test:test@#{Testcontainers.get_host()}:#{PostgresContainer.port(container)}/app_test"
```
Comment (jarlah) [inline lib/mix/tasks/testcontainers_test.ex:19]: building on an assumption that PostgresContainer uses "test" as username and password. Which holds true, but is an assumption.

**Comment by singlefleck** [comment]
Hey @jarlah,
Sorry about the review, I’m new to Testcontainers

**Comment by jarlah** [comment]
no stress :)

**Pull request merged**